### ISSUE TITLE
Allow a custom compilation command

### ISSUE THREAD

**Issue by justin808** [rails/webpacker#2640]
Context: Many applications have custom webpack build commands that are not easy to get retrofitted into the stale checking and assets:precompile of integrations of rails/webpacker.

Could we allow a setting in the `config/webpacker.yml` file to allow a custom compilation command?

`custom_webpack_compile`: "yarn run my_default_custom_webpack_command"
`test.custom_webpack_compile: "yarn run my_custom_webpack_command_to_build_test_bundles`
`production.custom_webpack_compile: "yarn run my_custom_webpack_command_to_build_prodution_bundles`

This command would be run with the RAILS_ENV and the NODE_ENV.

The use case for this is that many React on Rails installations have custom webpack setups. For React on Rails, two configuration values set the compiler command for test and production envs. Development webpack compilation is typically handled by the Procfile.dev.

Suppose that React on Rails install with a custom webpack config wants to leverage the compiler's stale checking. The command to get the webpack config currently is the `config/webpack/production.js` Potentially, the code in those files could be adjusted to run the current webpack configurations. Allowing for a custom command would be much more direct.

@gauravtiwari @jakeNiemiec If you're agreeable to this, I'd create a PR with this enhancement.

**Comment by justin808** [comment]
The compiler calls the bin/webpack script.

So maybe one just needs to replace that script?

https://github.com/rails/webpacker/blob/master/lib/webpacker/compiler.rb#L70-L74

```ruby
      stdout, stderr, status = Open3.capture3(
        webpack_env,
        "#{RbConfig.ruby} ./bin/webpack",
        chdir: File.expand_path(config.root_path)
      )
```

Let me know your thoughts on that.

**Comment by justin808** [comment]
The problem with replacing the installed binstub is that only a Ruby script works.

PR #2650 addresses that limitation.

**Comment by justin808** [comment]
@gauravtiwari @jakeNiemiec Another thought is that the `compile` option makes the invocation of the `bin/webpack` use the `RbConfig.ruby` but many times people run `bin/webpack --watch` and that won't use the special ruby version. This seems sort of problematic. Maybe we don't ever need to configure the ruby version in the `compiler.rb` file at all?

**Comment by justin808** [comment]
OK -- I see why this was done. Windows which I cannot test

https://github.com/rails/webpacker/pull/1286#issuecomment-367951995

```
gauravtiwari commented on Feb 22, 2018
Thanks for the update - the last bit is prefixing all the ./bin/webpack or dev server calls (if any) with RbConfig.ruby so it works on windows too.

Open3.capture3(webpack_env, "#{RbConfig.ruby} ./bin/webpack")
```

The consequence is that to make the bin/webpack script work in a simple manner is that one would have to provide a wrapper ruby script. 

Here is an example:

```ruby
#!/usr/bin/env ruby

ENV["RAILS_ENV"] ||= ENV["RACK_ENV"] || "development"
ENV["NODE_ENV"]  ||= "development"

require "pathname"
ENV["BUNDLE_GEMFILE"] ||= File.expand_path("../../Gemfile",
  Pathname.new(__FILE__).realpath)

require "bundler/setup"
require "webpacker"
require "webpacker/webpack_runner"

APP_ROOT = File.expand_path("..", __dir__)

puts "bin/webpack is running #{APP_ROOT}/bin/webpack.sh"
exec "#{APP_ROOT}/bin/webpack.sh"
```

and then `webpack.sh` can contain a normal shell script...

That seems very unnecessary.

**Comment by justin808** [comment]
@gauravtiwari @jakeNiemiec Can this issue get some consideration?

I'm working hard to minimize anything extra in https://github.com/shakacode/react_on_rails that overlaps with rails/webpacker.

Thus, the new webpack setup I'm recommending to allow for Server-Side Rendering is based on the client-side rails/webpacker configuration: https://github.com/shakacode/react_on_rails_tutorial_with_ssr_and_hmr_fast_refresh

However, some installations will be too complex to migrate to this.

Any objections to this small change? The workaround above is just too inconvenient.

**Comment by gauravtiwari** [comment]
Thanks @justin808 Apologies for not replying sooner. 

I am currently working on a new API for Webpacker 6.0 that's why I am bit hesitant on merging some PRs. I will share details soon.

**Comment by justin808** [comment]
@gauravtiwari Would you be open to a PR for this?

**Comment by justin808** [comment]
Fixed in #2650
